### PR TITLE
Update make, lots of linting

### PR DESCRIPTION
**Pull request by jthurman42**
Update to a more verbose linting and make config.

Big-ish changes:
* Moved the `cmd` package into `internal/cmd`
* Move `main.go` to `cmd/go-newrelic` (result is binary now compiled into `bin/<os>/go-newrelic`

Most other changes are linting related.  Lots of error checking on `w.Write()` (Should refactor these tests at a future date)

The `api` package has not been moved, so this *should* be backwards compatible with external users or the library, unless they are wrapping `cmd` which would seem like an odd thing to do.

Due to that, we *might* want to consider this a for a `5.0.0` release, and if so I'd argue for some more refactoring first (breaking out `api` into domain specific packages, moving under `pkg`, a bigger conversation to have)

Alternatively, we can rollback the move of `cmd` package and change the Makefile to deal with it.

Makefile enhancements:
* Run unit and Integration tests
* Auto-generate documentation
* Use golangci-lint for stricter linting
* Easier cross-compile, directory structure for building multiple platforms
* Travis config updated as well with ci specific make commands

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -1,13 +1,57 @@
...
-	printf "%s\n" "$(NOVENDOR)" | xargs -I {} sh -c 'golint -set_exit_status {}'
+# Determine commands by looking into cmd/*
+COMMANDS = $(wildcard ${SRCDIR}/cmd/*)
+
+GO_FILES := $(shell find $(COMMANDS) $(PACKAGES) -type f -name "*.go")
+
+# Determine binary names by stripping out the dir names
+BINS=$(foreach cmd,${COMMANDS},$(notdir ${cmd}))
+
+LDFLAGS='-X main.Version=$(PROJECT_VER)'
```
Comment (sanderblue) [inline Makefile:35]: Nice

File: Makefile [hunk 2]
```diff
@@ -1,13 +1,57 @@
...
+	@echo "=== $(PROJECT_NAME) === [ clean            ]: removing binaries and coverage file..."
+	@rm -rfv $(BUILD_DIR)/* $(COVERAGE_DIR)/*
+
+# Import fragments
+include build/deps.mk
+include build/compile.mk
+include build/testing.mk
+include build/package.mk
+include build/util.mk
+include build/document.mk
```
Comment (sanderblue) [inline Makefile:55]: 👍 on breaking these out into individual files

File: go.mod [hunk 3]
```diff
@@ -8,6 +8,12 @@ require (
 	github.com/imdario/mergo v0.3.8
 	github.com/olekukonko/tablewriter v0.0.2
 	github.com/spf13/cobra v0.0.5
+	github.com/spf13/pflag v1.0.5 // indirect
```
Comment (ctrombley) [inline go.mod:11]: The `tools-update` target is installing some test dependencies that end up introducing all these indirect references..  If I remember correctly, these caused us some merge issues early on.  

If I delete all the refs in `go.mod` after running `make` and then run a `go mod tidy`, they go away.  Is there a good way to automatically do this after the test run?

File: build/testing.mk [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+#
+
+
+
+lint: deps
+	@echo "=== $(PROJECT_NAME) === [ lint             ]: Validating source code running $(GOLINTER)..."
+	@$(GOLINTER) run ./...
+
+test: test-only
+test-only: test-unit test-integration
```
Comment (ctrombley) [inline build/testing.mk:12]: Can we add a line in the development section of the readme that show how to tag tests as unit or integration?

**Review by ctrombley — CHANGES_REQUESTED** [review]

**Comment by jthurman42** [comment]
This needs some serious rebase work, will reopen at another time.

**Pull request closed without merging**